src/procedures/proc_bfs.c

```diff
@@ -0,0 +1,222 @@
+/*
+* Copyright 2018-2020 Redis Labs Ltd. and Contributors
+*
+* This file is available under the Redis Labs Source Available License Agreement
+*/
+
+#include "proc_bfs.h"
+#include "../value.h"
+#include "../config.h"
+#include "../util/arr.h"
+#include "../query_ctx.h"
+#include "../util/rmalloc.h"
+#include "../datatypes/array.h"
+#include "../graph/graphcontext.h"
+#include "../algorithms/LAGraph_bfs_both.h"
+
+// MATCH (a:User {id: 1}) CALL algo.bfs(a, 0, 'MANAGES', true) YIELD nodes, edges
+typedef struct {
+	Graph *g;                       // Graph.
+	GrB_Index n;                    // Total number of results.
+	GrB_Index *node_ids;            // Array of connected nodes.
+	GrB_Vector parents;             // Vector associating each node in the BFS tree with its parent.
+	int reltype_id;                 // ID of relationship matrix to traverse.
+	int nodes_output_idx;           // Offset of nodes array in outputs
+	int edges_output_idx;           // Offset of edges array in outputs
+	SIValue *output;                // Array with a maximum of 4 entries: ["nodes", nodes, "edges", edges].
+	bool depleted;                  // True if BFS has already been performed for this node.
+} BFSContext;
+
+static ProcedureResult Proc_BFS_Invoke(ProcedureCtx *ctx, const SIValue *args) {
+	if(array_len((SIValue *)args) != 4) return PROCEDURE_ERR;
+	if(SI_TYPE(args[0]) != T_NODE                 ||   // Source node.
+	   SI_TYPE(args[1]) != T_INT64                ||   // Max level to iterate to, unlimited if 0.
+	   !(SI_TYPE(args[2]) & (T_NULL | T_STRING))  ||   // Relationship type to traverse if not NULL.
+	   SI_TYPE(args[3]) != T_BOOL)                     // Whether traversed edges should be returned.
+		return PROCEDURE_ERR;
+
+	BFSContext *pdata = ctx->privateData;
+	pdata->depleted = false;
+	Node *source_node = args[0].ptrval;
+	GrB_Index source_id = ENTITY_GET_ID(source_node);
+	int64_t max_level = args[1].longval;
+	/* The BFS algorithm uses a level of 1 to indicate the source node. If this value is not
+	 * zero (unlimited), increment it by 1 to make level 1 indicate the source's direct neighbors. */
+	if(max_level > 0) max_level++;
```

Comment:
The algorithm levels by default mean that 0 is unreachable, 1 is the source, and 2 are the direct neighbors. Our procedure here makes 1 the direct neighbors, since I feel like that's more intuitive.